### PR TITLE
Add functional priority queue Add FPriorityQueue<T>: immutable binary heap and tests

### PR DESCRIPTION
**Pull request by SeydaKinaci**
## Summary by CodeRabbit

- **New Features**
  - Introduced a new persistent priority queue that lets you add elements with an associated priority, view the highest priority item, remove it, and check the queue’s size.
- **Tests**
  - Added comprehensive unit tests to verify the correct ordering, proper behavior when the queue is empty, and overall robustness of the new functionality.

### CODE REVIEW

File: src/main/java/org/fungover/breeze/collection/FPriorityQueue.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    public FPriorityQueue<T> dequeue() {
+        if(heap.isEmpty()) return this;
+
+        List<Node<T>> newHeap = new ArrayList<>(heap);
+        Collections.swap(newHeap, 0, heap.size() - 1);
+        newHeap.remove(newHeap.size() - 1);
+
+        return new FPriorityQueue<>(heapifyDown(new ArrayList<>(newHeap),0));
```
Comment (liamr03) [inline src/main/java/org/fungover/breeze/collection/FPriorityQueue.java:60]: Optimization: Avoids unnecessary extra copy of the list, making dequeue() slightly faster.

return new FPriorityQueue<>(heapifyDown(newHeap, 0));

File: src/main/java/org/fungover/breeze/collection/FPriorityQueue.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+            if(right<size && heap.get(right).priority < heap.get(smallest).priority){
+                smallest = right;
+            }
+            if(smallest == index)
+                break;
+
+            Collections.swap(heap, index, smallest);
+            index = smallest;
+        }
+        return heap;
```
Comment (liamr03) [inline src/main/java/org/fungover/breeze/collection/FPriorityQueue.java:100]: Line 100:
Optimization: If no swaps were made, return originalHeap to avoid an unnecessary copy.

return swapped ? heap : originalHeap;

File: src/test/java/collection/FPriortiyQueueTest.java
Comment (liamr03) [inline src/test/java/collection/FPriortiyQueueTest.java:1]: Maybe an improvement would be a test case for handling duplicate priorities.
Right now, all elements have unique priorities. What if multiple elements have the same priority? Example:
    ("ItemA", 1);
    ("ItemB", 1);
    ("ItemC", 1);

**Review by liamr03 — COMMENTED** [review]
Very well written code, there is not any critical errors I can see, I have written some improvements/optimizations you can implement if you want :)

**Review by liamr03 — DISMISSED** [review]

File: src/main/java/org/fungover/breeze/collection/FPriorityQueue.java (hunk 1 above)
Comment (SeydaKinaci) [reply src/main/java/org/fungover/breeze/collection/FPriorityQueue.java]: Thanks for the suggestion @liamr03! It’s definitely better this way. I fixed it. :)

File: src/main/java/org/fungover/breeze/collection/FPriorityQueue.java (hunk 2 above)
Comment (SeydaKinaci) [reply src/main/java/org/fungover/breeze/collection/FPriorityQueue.java]: That’s a great point! I fixed it. Thanks @liamr03 ! :)

File: src/test/java/collection/FPriortiyQueueTest.java
Comment (SeydaKinaci) [reply src/test/java/collection/FPriortiyQueueTest.java:1]: I didn’t think of that. Good point:) I can add it. Thanks!

**Review by liamr03 — DISMISSED** [review]
Looks good!

**Review by voffie — COMMENTED** [review]
Just saw that the commits aren't verified. This needs to be fixed before it can be merged. @kappsegla can help guide you in the right direction

**Comment by SeydaKinaci** [comment]
I thought you might be able to review it again since you've already looked at my code before. If you have time, I would really appreciate it if you could check it again @liamr03 @voffie :) I've signed my commits and completed the missing parts.

File: src/main/java/org/fungover/breeze/collection/FPriorityQueue.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    public FPriorityQueue<T> dequeue() {
+        if(heap.isEmpty()) return this;
+
+        List<Node<T>> newHeap = new ArrayList<>(heap);
+        Collections.swap(newHeap, 0, heap.size() - 1);
+        newHeap.remove(newHeap.size() - 1);
+
+        return new FPriorityQueue<>(heapifyDown(newHeap, 0));
+    }
+
```
[the remark this answers is not in the document]
Comment (Lhlmlund) [reply src/main/java/org/fungover/breeze/collection/FPriorityQueue.java:62]: also can remove  extra  new arraylist from 

return new FPriorityQueue<>(heapifyDown(new ArrayList<>(newHeap), 0));


instead keep : 
return new FPriorityQueue<>(heapifyDown(newHeap, 0));

File: src/main/java/org/fungover/breeze/collection/FPriorityQueue.java [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    /**
+     * Constructs a new priority queue with a given heap.
+     * This constructor is used internally to create new instances while maintaining immutability.
+     *
+     * @param heap The heap to be used for this priority queue.
+     */
+    private FPriorityQueue(List<Node<T>> heap) {
+        this.heap = List.copyOf(heap);
+    }
+
```
Comment (Lhlmlund) [inline src/main/java/org/fungover/breeze/collection/FPriorityQueue.java:36]: this constructor creates an unmodifiable list when it should work with mutable data, it makes it unnecessary since you make a fresh copy in enqueue()
   
   change this:
   
    private FPriorityQueue(List<Node<T>> heap) {
        this.heap = List.copyOf(heap);
    }
    
    to this 
    
    private FPriorityQueue(List<Node<T>> heap) {
    this.heap = heap; // Keep the mutable copy
}

**Review by Lhlmlund — CHANGES_REQUESTED** [review]
Your FPriorityQueue implementation is solid overall but needs a few performance changes, ive made some comments that you could take a look at

**Review by JenDevs — CHANGES_REQUESTED** [review]

**Review by JenDevs — DISMISSED** [review]
Tests run as expected and the implementation looks good. I noticed a few edge cases (equal priorities and some heapify paths) aren’t covered yet, but overall I approve the changes.

File: pom.xml
Comment (voffie) [inline pom.xml:1]: I believe this change is committed by mistake. Make sure to check what changes are being made. Due to the CODEOWNERS file, these types of changes are forcing a review from the admin-team

File: src/main/java/org/fungover/breeze/collection/FPriorityQueue.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+     * @return A new FPriorityQueue instance with the element inserted in sorted order.
+     */
+    public FPriorityQueue<T> enqueue(T element, int priority ) {
+        List<Node<T>> newHeap = new ArrayList<>(heap);
+        newHeap.add(new Node<>(element, priority));
+        heapifyUp(newHeap, newHeap.size() -1);
+        return new FPriorityQueue<>(newHeap);
+    }
+
+    public FPriorityQueue<T> dequeue() {
```
Comment (voffie) [inline src/main/java/org/fungover/breeze/collection/FPriorityQueue.java:53]: This function is missing Javadoc

File: src/main/java/org/fungover/breeze/collection/FPriorityQueue.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+                new FPriorityQueue<>();
+
+        List<Node<T>> newHeap = new ArrayList<>(heap);
+        Collections.swap(newHeap, 0, heap.size() - 1);
+        newHeap.remove(newHeap.size() - 1);
+
+        return new FPriorityQueue<>(heapifyDown(newHeap, 0));
+    }
+
+    public T peek(){
```
Comment (voffie) [inline src/main/java/org/fungover/breeze/collection/FPriorityQueue.java:64]: This function is missing Javadoc

File: src/main/java/org/fungover/breeze/collection/FPriorityQueue.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+       if(isEmpty())
+        throw new NoSuchElementException("Priority queue is empty");
+       return heap.get(0).element;
+    }
+
+    public boolean isEmpty(){
```
Comment (voffie) [inline src/main/java/org/fungover/breeze/collection/FPriorityQueue.java:70]: This function is missing Javadoc

File: src/main/java/org/fungover/breeze/collection/FPriorityQueue.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        return heap.isEmpty();
+    }
+
+    public int size() {
```
Comment (voffie) [inline src/main/java/org/fungover/breeze/collection/FPriorityQueue.java:74]: This function is missing Javadoc

File: src/main/java/org/fungover/breeze/collection/FPriorityQueue.java [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        return heap.size();
+    }
+
+    public static <T> List<Node<T>> heapifyDown(List<Node<T>> originalHeap, int index) {
```
Comment (voffie) [inline src/main/java/org/fungover/breeze/collection/FPriorityQueue.java:78]: This function is missing Javadoc

**Review by voffie — DISMISSED** [review]
The code looks good! There are just a few things to look out for in the future!

**Review by voffie — APPROVED** [review]
Looks good! Well done :)

**Review by JenDevs — APPROVED** [review]
Looking good. Well done!
